api/src/test/java/org/apache/iceberg/TestPartitionSpecValidation.java

```diff
@@ -123,6 +123,21 @@ public void testMultipleDatePartitionsWithDifferentSourceColumns() {
     PartitionSpec.builderFor(SCHEMA).hour("d").hour("another_d").build();
   }
 
+  @Test
+  public void testMultipleIdentityPartitions() {
+    PartitionSpec.builderFor(SCHEMA).year("d").identity("id").identity("d").identity("s").build();
```

Comment:
I think I understand the goal this PR tires to achieve, but what is the use case for this? Because they are now not deduped, there are now 2 partition columns "d_year" and "d". Is this actually the right behavior for users?

Reply:
I was having a similar question but forgot to ask, not related to this change - why do we not want to have two different time transforms on the same column? Is it just because the more granular one makes the other one not so useful (e.g. when already partitioned by `day("d")`, partition on `year("d")` won't help with query performance)?

Reply:
Users might not define their spec like this test, which checks one possible code path, where dudup won't apply for date and identity and also won't apply if identity partitions are for different source fields. 

Now, partition spec update can actually add day and year for the same source field, we may consider to drop this constraint. @rdblue how do you feel about this?

Reply:
This check was introduced because users were accustomed to partitioning by both date and hour, assuming that hour was in [0-23]. In Iceberg, `hours` is the granularity, not a repeating value. We had people using `day("ts").hour("ts")` often, which is redundant so we want to fail cases like that because it signals the user does not understand what they are using.

When updating partitioning, we allow duplication because there is probably existing data in the table stored by day. If the user had to drop `ts_day` to add `ts_hour`, then the metadata tables would also be affected. The result is that metadata queries to find `max(partition.ts_day)` would start to fail. Also, there is no harm in populating `ts_day` for data partitioned by `ts_hour` so we allow it.

Reply:
> We had people using day("ts").hour("ts") often, which is redundant so we want to fail cases like that because it signals the user does not understand what they are using.

Yes people also asked me about this, and there is little information about this when people read https://iceberg.apache.org/partitioning. Let me add some more doc in that area, plus the info about `UpdatePartitionSepc` and its SQL extension.